src/core/algorithm/timelinePagerank.test.js

```diff
@@ -165,5 +170,38 @@ describe("src/core/algorithm/timelinePagerank", () => {
       // src has the weight, and dst doesnt, so it should have a higher score
       expect(srcScore).toBeGreaterThan(dstScore);
     });
+    it("satisfies the flow conservation invariants", async () => {
+      const {nodeOrder, edgeOrder, result, graph} = await example();
+      const {
+        distribution,
+        forwardFlow,
+        backwardFlow,
+        syntheticLoopFlow,
+        seedFlow,
+      } = result;
+      // For any node: Its score is equal to the sum of:
+      // - The score it received from the seed vector
+      // - The score it received from every in edge
+      // - The score it received from every out edge
+      // - The score it received from its self loop
+      const nodeToExpectedScore = new Map();
+      function addScore(a: NodeAddressT, score: number) {
+        const existing = nodeToExpectedScore.get(a) || 0;
+        nodeToExpectedScore.set(a, existing + score);
+      }
+      nodeOrder.forEach((na, i) => {
+        addScore(na, seedFlow[i] + syntheticLoopFlow[i]);
+      });
+      edgeOrder.forEach((ea, i) => {
+        const {src, dst} = NullUtil.get(graph.edge(ea));
+        addScore(src, backwardFlow[i]);
+        addScore(dst, forwardFlow[i]);
+      });
+      nodeOrder.forEach((na, i) => {
+        const expected = nodeToExpectedScore.get(na);
+        const actual = distribution[i];
+        expect(expected).toBeCloseTo(actual);
+      });
```

Comment:
The default number of digits for `toBeCloseTo` is just 2. Which seems very coarse to handle rounding errors.
Also error output wouldn't be very helpful here as there's no address for context, and it will stop at the first error.

Ideally, the error would show a list of `{address, expected, actual}` where the difference is outside of tolerance.

Reply:
Sure, bumped it up to 4 digits.